Documentation update for avvio PR 82 (#1498)
fastify/fastify: docs/Plugins.md

})
```

> Functions in the `options` parameter are also supported and will be evaluated at the time the plugin is registered while giving access to the fastify instance via the first positional argument:

Reviewer: ```suggestion
The `options` parameter can also be a `Function` which will be evaluated at the time the plugin is registered while giving access to the fastify instance via the first positional argument:
```

```js

> fastify.register((fastify, opts, next) => {

Reviewer: I think this should be encapsulated in `fastify-plugin` to work. Maybe you could also add a couple of unit tests for this functionality, so we check we don't regress here for any reason.

Author: I will get to creating tests (as well as the other comments) in the weekend as I was in transit today, sorry for the late response.

Author: Hmm it appears that the `avvio` update I did does not work correctly on fastify. Even though the function as an option is called, the variable passed to it is not the external fastify reference. Probably has something to do with how `override` works in `fastify.js`? I tried debugging it to no avail as I could not find `avvio.use` being called anywhere, only `middie`. 

You can try the below simplified snippet on `plugin.test.js` to verify:
```js
test('fastify.register with fastify-plugin should provide access to external fastify instance if opts argument is a function', t => {
  t.plan(13)
  const fastify = Fastify()

  fastify.register((instance, opts, next) => {
    instance.register(fp((i, o, n) => {
      i.decorate('test', () => {})
      t.ok(i.test)
      n()
    }))

    // Below line will fail
    instance.register(fp((i, o, n) => n()), p => t.ok(p.test))

    // Will succeed but is expected (?)
    instance.register((i, o, n) => n(), p => t.notOk(p.test))

    // Once again, fails
    instance.register((i, o, n) => n(), p => t.ok(p === instance || p === fastify))

    t.notOk(instance.test)

    // the decoration is added at the end
    instance.after(() => {
      t.ok(instance.test)
    })

    instance.get('/', (req, reply) => {
      t.ok(instance.test)
      reply.send({ hello: 'world' })
    })

    next()
  })

  fastify.ready(() => {
    t.notOk(fastify.test)
  })

  fastify.listen(0, err => {
    t.error(err)
    fastify.server.unref()

    sget({
      method: 'GET',
      url: 'http://localhost:' + fastify.server.address().port
    }, (err, response, body) => {
      t.error(err)
      t.strictEqual(response.statusCode, 200)
      t.strictEqual(response.headers['content-length'], '' + body.length)
      t.deepEqual(JSON.parse(body), { hello: 'world' })
    })
  })
})
```

Reviewer: I think this is a bug in avvio, not in Fastify. Try reproducing it there first, you should be able to. You want to invoke the function not when the plugin is registered, but before the plugin is executed.

Author: A test case already covers for that on `avvio`, if it was evaluated on registration L170 would fail on [this test](https://github.com/mcollina/avvio/blob/master/test/basic.test.js#L150). Plus, L159 succeeds whereas in the above `fastify` test it fails. I firmly believe it has to do with the `override` that is being replaced, let me know your thoughts.

Reviewer: This use of override is correct, and works as intended. We should probably add some tests about override in avvio. That test does not set the property on the ‘s’ instance, but on the outside one.

Author: You are correct, I made some changes to `avvio` as the test case was indeed changing the outer variable thus rendering it invalid. I have created a new PR.

The `parent` we were binding to the function passed in as `options` was actually the parent variable, not the `server` instance. Instead of binding, I made it evaluate the `options` during `exec` as it did before but this time with `this.server` being passed in and **after** `override` is called. You can view the PR [here](https://github.com/mcollina/avvio/pull/83).

Reviewer: Can you add that test now that it was fixed in avvio?

  fastify.decorate('foo_bar', { hello: 'world' })

  next()
})

// The opts argument of fastify-foo will be { hello: 'world' }
fastify.register(require('fastify-foo'), parent => parent.foo_bar)
```

> The fastify instance passed on to the function is the latest state of the **global fastify instance** allowing access to variables injected via `decorate` by preceding plugins according to the **order of registration**. This is useful in case a plugin depends on changes made to the Fastify instance by a preceding plugin f.e. utilizing an existing database connection to wrap around it.

Reviewer: This still mention "global".

Author: Resolved in latest commit.
